File: Leetcode 1429 First Unique Number [medium] [design]/First Unique Number [design].java
class FirstUnique {
    private Queue<Integer> queue = new LinkedList<>();
    private Map<Integer, Integer> map = new HashMap<>();
    
    public FirstUnique(int[] nums) {
        for (int num : nums) {
            add(num);
        }
    }
    
    public int showFirstUnique() {
        while (!queue.isEmpty() && map.get(queue.peek()) > 1) {
            queue.poll();
        }
        
        if (queue.isEmpty()) {
            return -1;
        }
        
        return queue.peek();
    }
    
    public void add(int value) {
        if (!map.containsKey(value)) {
            map.put(value, 1);
        } else {
            // map裡面調用put(), 不能用++, 必須用 val + 1
            map.put(value, map.get(value) + 1);
        }
        queue.offer(value);
    }
}

/**
 * Your FirstUnique object will be instantiated and called as such:
 * FirstUnique obj = new FirstUnique(nums);
 * int param_1 = obj.showFirstUnique();
 * obj.add(value);
 */

File: Leetcode 1429 First Unique Number [medium] [design]/First Unique Number [design].js
/**
 * @param {number[]} nums
 */
var FirstUnique = function(nums) {
    this.queue = [];
    this.map = new Map();
    for (let num of nums) {
        this.add(num);
    }
};

/**
 * @return {number}
 */
FirstUnique.prototype.showFirstUnique = function() {
    while (this.queue.length > 0 && this.map.get(this.queue[0]) > 1) {
        this.queue.shift();
    }
    
    if (this.queue.length === 0) {
        return -1;
    }
    
    return this.queue[0];
};

/** 
 * @param {number} value
 * @return {void}
 */
FirstUnique.prototype.add = function(value) {
    if (!this.map.has(value)) {
        this.map.set(value, 1);
    } else {
        this.map.set(value, this.map.get(value) + 1);
    }
    this.queue.push(value);
};

/** 
 * Your FirstUnique object will be instantiated and called as such:
 * var obj = new FirstUnique(nums)
 * var param_1 = obj.showFirstUnique()
 * obj.add(value)
 */